tfrecord/torch/dataset.py

```diff
@@ -164,16 +165,38 @@ def __init__(self,
 
     def __iter__(self):
         worker_info = torch.utils.data.get_worker_info()
+
         if worker_info is not None:
             np.random.seed(worker_info.seed % np.iinfo(np.uint32).max)
+
+            # Each worker will get only len(splits) // gcd(len(splits), num_workers) files
+            # and each such file will be sharded among num_workers // gcd(len(splits, num_workers)) workers
+            gcd = math.gcd(len(self.splits), worker_info.num_workers)
+            worker_num_splits = len(self.splits) // gcd
+            start = (worker_info.id % gcd) * worker_num_splits
+            end = start + worker_num_splits
+
+            worker_splits = dict(self.splits[start:end])
+
+            shard = worker_info.id // gcd, worker_info.num_workers // gcd
+
```

Comment:
This is not right. This won't read all of the data. I suggest before sending a PR to at least create one test example and make sure that it works.

Reply:
Well, the idea was, that if you have i.e. 12 files and 3 workers, you could either give all workers one third of each file or give each worker on third of the files - so first worker would get first 4 files, etc., each of the file unsharded.

The reasoning for choosing this implementation was to reduce the strain on file system as it reduces number of opened files by each worker. Sadly, I'm working with quite slow filesystem, so I can actually see the benefits of using less files by each process.

It has a small drawback in not keeping the ratios exactly the same, but only the within one worker.

I've actually tested it on my files and it seems to work properly - all data are read after one run if `infinite=False` and there is less opened files.

Reply:
If you're not comfortable with this implementation, we can of course do it in a more straightforward way by giving all the workers all the files and a "regular" shard (`(worker_info.id, worker_info.num_workers)`) and then of course also deleting the conversion of splits into an ordered list of tuples and back.

But I'd argue this brings a benefit to users with slower file system as it lowers the amount of opened files and has the basically same results.